zun/container/k8s/driver.py

```diff
@@ -215,208 +82,204 @@ def __init__(self):
         self.custom = client.CustomObjectsApi()
         self.net_v1 = client.NetworkingV1Api()
 
+        # self.network_api = zun_network.api(admin_context, self.net_v1)
+        k8s_network = network.K8sNetwork()
+        k8s_network.init(admin_context, self.net_v1)
+        self.network_driver = k8s_network
+
+        self.volume_driver = volume.K8sConfigMap(self.core_v1)
+
+        utils.spawn_n(self._watch_pods, admin_context)
+
+    def _watch_pods(self, context):
+        def _do_watch():
+            watcher = watch.Watch()
+            for event in watcher.stream(
+                self.core_v1.list_pod_for_all_namespaces,
+                label_selector=f"{mapping.LABELS['type']}=container"):
+                pod = event["object"]
+                event_type = event["type"]
+                container_uuid = pod.metadata.labels[mapping.LABELS["uuid"]]
+                try:
+                    container = objects.Container.get_by_uuid(context, container_uuid)
+                    self._sync_container(container, pod, pod_event=event_type)
+                    container.save()
+                    LOG.info(
+                        f"Synced {container_uuid} to k8s state after {event_type} event")
+                except exception.ContainerNotFound:
+                    # Just skip this pod, it should be cleaned up on the periodic sync
+                    LOG.info(f"Skipping update on missing container '{container_uuid}'")
+
+        backoff, max_backoff = 0.0, 128.0
+        def _get_backoff(current, maximum):
+            return min(max(current * 2, 1.0), maximum)
+
+        while True:
+            try:
+                if backoff:
+                    time.sleep(backoff)
+                _do_watch()
+            except client.ApiException as exc:
+                LOG.error(f"Unexpected K8s API error: {exc}")
+                backoff = _get_backoff(backoff, max_backoff)
+            except Exception as exc:
+                # This indicates a business logic failure; our code is wrong. Keep
+                # the loop going but log the exception; possibly future watch events
+                # will not always trigger this error and we can keep the cluster from
+                # getting too far out of sync.
+                LOG.exception("Unexpected error watching pods")
+                backoff = _get_backoff(backoff, max_backoff)
+
     def periodic_sync(self, context):
         """Called by the compute manager periodically.
 
         Use this to perform cleanup in case state on K8s has diverged from desired.
         """
-        # Ensure all namespaces have a default network policy
+        # TODO(jason): delete dangling neutron ports
```

Comment:
Just calling your attention to some of the TODOs that are left in there. I think they are mostly related to cleanup.